### PR TITLE
Fizzy apply

### LINKED ISSUE #112: Support for fuzzy patches
When generating a unified diff with context, I would like to be able to do a fuzzy apply of that patch. Does your library support this?

For example, take the following unified diff with context 3:

```
--- original
+++ modified
@@ -2,6 +2,9 @@
 2
 3
 4
+a
+b
+c
 5
 6
 7
```

And my file looks like the following (exactly the original but with three lines added to the top):

```
x
y
z
1
2
3
4
5
6
7
8
```

Using the linux `patch` tool I am still able to apply this patch:

```
$ patch -o - neworig udiff.patch
patching file - (read from neworig)
Hunk #1 succeeded at 5 with fuzz 1 (offset 3 lines).
```

When using `DiffUtils.patch` of your library I get an error (because the line numbers no longer match):

```
Incorrect Chunk: the chunk content doesn't match the target
```

Maybe there is a configuration option to enable fuzzy patching?

### PR DESCRIPTION
**Pull request by anatawa12**
Closes #112

### CODE REVIEW

File: java-diff-utils/src/main/java/com/github/difflib/patch/AbstractDelta.java [hunk 1]
```diff
@@ -70,6 +70,18 @@ protected VerifyChunk verifyAntApplyTo(List<T> target) throws PatchFailedExcepti
     
     protected abstract void restore(List<T> target);
     
+    /**
+     * Apply patch fuzzy.
+     *
+     * @param target the list this patch will be applied to
+     * @param fuzz the count of elements which ignores prefix and suffix
```
Comment (cowwoc) [inline java-diff-utils/src/main/java/com/github/difflib/patch/AbstractDelta.java:77]: I have a hard time understanding this parameter description. Please try rephrasing it and/or embed an example in the Javadoc.

**Comment by wumpz** [comment]
Sorry. Fortgot about this patch. I will look into it.

**Comment by wumpz** [comment]
Could you first resolve the conflicts? Strangely I am not able to resolve it using Github.

**Comment by anatawa12** [comment]
> Could you first resolve the conflicts? Strangely I am not able to resolve it using Github.

Yes. i'll resolve conflicts and add more documentation about fuzzy patch.

File: java-diff-utils/src/main/java/com/github/difflib/patch/AbstractDelta.java [hunk 2]
```diff
@@ -74,7 +74,10 @@ protected VerifyChunk verifyAntApplyTo(List<T> target) throws PatchFailedExcepti
      * Apply patch fuzzy.
      *
      * @param target the list this patch will be applied to
-     * @param fuzz the count of elements which ignores prefix and suffix
+     * @param fuzz the count of first/end elements which is not matched.
```
Comment (cowwoc) [inline java-diff-utils/src/main/java/com/github/difflib/patch/AbstractDelta.java:77]: I didn't understand the concept of fuzzy patch before reading this Javadoc and I was unable to understand it. I don't think this is your fault. I don't think we'll be able to explain the parameters unless the user understands the purpose/meaning of `fuzzy patch` to begin with.

I suggest replacing the parameter Javadoc with something along the lines of: `@param fuzz the number of elements to ignore before/after the patched elements"` and adding `@see <a href="https://www.gnu.org/software/diffutils/manual/html_node/Inexact.html">Description of Fuzzy Patch</a> for more information`.

**Comment by cowwoc** [comment]
The new documentation look good to me. Thank you!

**Comment by wumpz** [comment]
Merged. Thx for the improvement. Could this something for the Wiki documentation?

**Comment by meteorcloudy** [comment]
We are using applyFuzzy in Bazel (see https://github.com/bazelbuild/bazel/issues/17897), but it's leading to a problem of applying a clean patch with multiple chunks: https://github.com/kythe/kythe/issues/5864

My debugging shows there is something wrong about calculating the `defaultPosition` at https://github.com/java-diff-utils/java-diff-utils/blob/281c073c56a0cd0052963562e007c51ce2994983/java-diff-utils/src/main/java/com/github/difflib/patch/Patch.java#L108-L115

Do I understand correctly that `lastPatchDelta` means the offset we should apply to the next delta after applying the previous one? If so it probably should not be `lastPatchDelta = patchPosition - delta.getSource().getPosition();` since the new chunk could be smaller.

The problematic patch I encountered is:

```
--- src/include/souffle/utility/FileUtil.h
+++ src/include/souffle/utility/FileUtil.h
@@ -20,23 +20,16 @@
 #include <climits>
 #include <cstdio>
 #include <cstdlib>
-#include <filesystem>
 #include <fstream>
 #include <map>
-#include <optional>
 #include <sstream>
 #include <string>
 #include <utility>
 #include <sys/stat.h>

-// -------------------------------------------------------------------------------
-//                               File Utils
-// -------------------------------------------------------------------------------
-
 #ifndef _WIN32
 #include <unistd.h>
 #else
-#define NOMINMAX
 #define NOGDI
 #include <fcntl.h>
 #include <io.h>
@@ -44,11 +37,21 @@
 #include <windows.h>

 // -------------------------------------------------------------------------------
-//                               Windows
+//                               File Utils
 // -------------------------------------------------------------------------------

+#define X_OK 1 /* execute permission - unsupported in windows*/
+
 #define PATH_MAX 260

+/**
+ * access and realpath are missing on windows, we use their windows equivalents
+ * as work-arounds.
+ */
+inline int access(const char* path, int mode) {
+    return _access(path, mode);
+}
+
 inline char* realpath(const char* path, char* resolved_path) {
     return _fullpath(resolved_path, path, PATH_MAX);
 }
```

**Pull request merged**